### DISCUSSION TITLE
Loading a sqlite extension

### DISCUSSION THREAD

**Discussion by davidmichallet** [sqlpage/SQLPage#557] [Q&A]
Good morning,
Many thanks for this wonderful framework and all its possibilities...🎉🎊🎉🎊🎉
It's beautiful and easy to create a powerfull app !

Member of a small consumer association, I am in the process of setting up a platform for managing orders, products, etc. (I will share the code when it is more advanced)

I'm running into a problem, I need to create a (dynamic) summary table where I'm transposing rows into columns.
I use "dbeaver" for management/debugging of my database and when I run the code below, everything works!
I use sqlite database.
sqlpage v0.26

```sql
with columns as (
  select distinct pro.reference as reference, pro.name
  from pro product
  LEFT JOIN commandline lc ON pro.reference = lc.reference 
  WHERE lc.quantity > 0 AND lc.number = 7
),
lines as (
    select 'drop view if exists v_synthese_distrib_commande_num_; 'as part
  union all
  select 'create view v_synthese_distrib_commande_num_ as '
  union all
  select 'select adh.name '
  union all
  select ', sum(quantity) filter (where lc.reference = ' || reference || ') as "' || name || '" '
  from columns
  UNION ALL
  select ', sum(quantity) AS Total '
  union all
  select 'FROM order_line lc LEFT JOIN adherent adh ON lc.id_adherent = adh.id_adherent LEFT JOIN product p ON lc.reference = p.reference WHERE lc.quantity > 0 AND lc.numero = 7 group by lc.id_adherent order by lc. member_id;'
)
select eval(group_concat(part,'')) from lines;
```

and when I play it with SQLPAGE, it does not recognize the "eval" function?

```
error returned from database: (code: 1) no such function: eval (at statement byte offset 793)

(code: 1) no such function: eval (at statement byte offset 793)
```
However I tried to put the "define" extension in the sqlpage.json

```json
{
  "database_url": "sqlite://./sqlpage/sqlpage.db?mode=rwc",
    "max_uploaded_file_size": 200000,
      "port": "8080",
  "allow_exec":true,
  "sqlite_extensions": "mod_spatialite crypto define regex eval"
}
```

Do you have an idea?

Thank you in advance!

**Comment by lovasoa** [comment]
Hello and welcome to SQLPage !
  In order to load sqlite extensions, you have to download them and make them loadable as libraries on your system. Then you have to load them at the start of sqlpage using

```json
{
  "database_url": "sqlite://./sqlpage/sqlpage.db?mode=rwc",
  "max_uploaded_file_size": 200000,
  "port": "8080",
  "allow_exec":true,
  "sqlite_extensions": ["mod_spatialite", "crypto", "define", "regex", "eval"]
}
```

Notice that the sqlite_extension object is an array, not a string.
If you use a string instead as mentioned in your message above, you should see the following in sqlpage's output in your terminal:

```logs
[2024-08-28T08:39:59.297Z ERROR sqlpage] Unable to load configuration
    
    Caused by:
        invalid type: string "mod_spatialite crypto define regex eval", expected a sequence
```

If you have specified the extensions list correctly but the extension cannot be found on your system, you will see the following warning messages repeated on the console until you add the extension:

```
[2024-08-28T08:33:41.927Z WARN  sqlpage::webserver::database::connect] Failed to connect to the database: error returned from database: (code: 0) eval.so: cannot open shared object file: No such file or directory. Retrying in 5 seconds.
```

If you placed the extension file in a non-standard library path for your system, you will need to specify the path to the extension file. For instance, you can decide to place it inside the `sqlpage` configuration directory in the current folder, and include it like so:

```json
{
  "database_url": "sqlite://./sqlpage/sqlpage.db?mode=rwc",
  "sqlite_extensions": ["./sqlpage/eval.so"]
}
```

**Comment by davidmichallet** [comment]
Thank you very much for your responsiveness,
the extension is present on my OS (linux), I put it in a subdirectory of sqlpage and despite your advice I can't play my query.
I test with sqlpage v0.27

I get the same error message on the web page and no message in the terminal ?

```json
{
  "database_url": "sqlite://./sqlpage/sqlpage.db?mode=rwc",
    "max_uploaded_file_size": 200000,
      "port": "8080",
  "allow_exec":true,
  "sqlite_extensions": ["./sqlpage/extensions/eval/eval.so"]
}
```

Error We are sorry, but an error occurred while generating this page. You should contact the site's administrator.

error returned from database: (code: 1) no such function: eval (at statement byte offset 793) Error in query: WITH columns AS (SELECT DISTINCT pro.reference AS reference, pro.name FROM product AS pro LEFT JOIN order_line AS lc ON pro.reference = lc.reference WHERE lc.quantite > 0 AND lc.numero = 7), lines AS (SELECT 'drop view if exists v_synthese_distrib_ command_num_; ' AS part UNION ALL SELECT 'create view v_synthese_distrib_command_num_ as ' UNION ALL SELECT 'select adh.name ' UNION ALL SELECT ', sum(quantity) filter (where lc.reference = ' || reference || ') as "' || name || '" ' FROM columns UNION ALL SELECT ', sum(quantity) AS Total ' UNION ALL S ELECT 'FROM lc_command line LEFT JOIN adherent adh ON lc.id_adherent = adh.id_adherent LEFT JOIN product p ON lc.reference = p.reference WHERE lc.quantity > 0 AND lc.numero = 7 group by lc.id_adherent order by lc.id_adherent;') SELECT eval(group_concat(part, '')) FROM lines;
 ⬆️ line 1, character 793 

Backtrace (code: 1) no such function: eval (at statement byte offset 793)

**Reply by lovasoa** [reply]
If you did not notice sqlpage refusing to start initially with the incorrect `sqlpage.json` file you posted earlier, then it means your configuration file is not being taken into account at all. Maybe you named it incorrectly, put it in the wrong folder, started sqlpage from the wrong folder, or did not restart sqlpage at all after the configuration file change ?

**Reply by lovasoa** [reply]
You can make sqlpage display its full configuration, as well as the path it reads it from when it starts, by setting the environment variable `RUST_LOG` to `sqlpage=debug `.

**Comment by davidmichallet** [comment]
if I add this piece of code at the start of my sql query:

`select load_extension('./sqlpage/eval.so');` 

I have this error message:

`Error We are sorry, but an error occurred while generating this page. You should contact the site's administrator.

error returned from database: (code: 1) not authorized Backtrace (code: 1) not authorized`

Thanks

**Reply by lovasoa** [reply]
See my response above. The extension should be loaded by sqlpage itself, not from your sql code, but it looks like your configuration file is not being read by sqlpage. Maybe it's in the wrong place ?

**Comment by davidmichallet** [comment]
Sorry 🙏,

You right, the "sqlpage.json" is not locate in the good directory...
Thanks a lot and have a nice day !
